### DISCUSSION TITLE
Could not create another Distribution certificate, reached the maximum number of available Distribution certificates.

### DISCUSSION THREAD

**Discussion by vin-fandemand** [fastlane/fastlane#19602] [Q&A]
### New Issue Checklist

- [ ] Updated fastlane to the latest version
- [x] I read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I read [docs.fastlane.tools](https://docs.fastlane.tools)
- [x] I searched for [existing GitHub issues](https://github.com/fastlane/fastlane/issues)

### Issue Description
I use `fastlane match` to create a provisioning profile for my apps. Whenever I need to create a profile or whenever I have to add a new test device, I use the `match` command to regenerate the profiles. However, I often run into the `Could not create another Distribution certificate, reached the maximum number of available Distribution certificates.` issue because of each time `match` keeps adding more Distribution certificates. Then I will have to revoke some older certficates, which ends up revoking some provisional profiles in use and then have to create them all over again. I tried with `fastlane match adhoc --generate_apple_certs=false` as well but really didnt work. Is there a way better way of handling this ?
##### Command executed
`fastlane match adhoc --generate_apple_certs=false`

##### Complete output when running fastlane, including the stack trace and command used

<details>
  <pre> [11:13:36]: fastlane detected a Gemfile in the current directory
[11:13:36]: However, it seems like you didn't use `bundle exec`
[11:13:36]: To launch fastlane faster, please use
[11:13:36]: 
[11:13:36]: $ bundle exec fastlane match adhoc --generate_apple_certs=false
[11:13:36]: 
[11:13:36]: Get started using a Gemfile for fastlane https://docs.fastlane.tools/getting-started/ios/setup/#use-a-gemfile
[11:13:38]: Successfully loaded '/Users/vineethpalani/Documents/Github/vio_server_panel-1/ios/fastlane/Matchfile' 📄

+--------------+---------------------------------------------+
|             Detected Values from './Matchfile'             |
+--------------+---------------------------------------------+
| git_url      | git@github.com:repo/qcertificates.git |
| git_branch   | viomsp-staging                             |
| storage_mode | git                                         |
| type         | adhoc                                       |
| username     | vineeth@fandemand.ca                        |
+--------------+---------------------------------------------+

+--------------------------------+---------------------------------------------+
|                          Summary for match 2.192.0                           |
+--------------------------------+---------------------------------------------+
| generate_apple_certs           | false                                       |
| type                           | adhoc                                       |
| readonly                       | false                                       |
| skip_provisioning_profiles     | false                                       |
| app_identifier                 | ["viomsp.ios.dev"]                         |
| username                       | vineeth@fandemand.ca                        |
| storage_mode                   | git                                         |
| git_url                        | git@github.com:repo/qcertificates.git |
| git_branch                     | viomsp-staging                             |
| shallow_clone                  | false                                       |
| clone_branch_directly          | false                                       |
| keychain_name                  | login.keychain                              |
| force                          | false                                       |
| force_for_new_devices          | false                                       |
| skip_confirmation              | false                                       |
| skip_docs                      | false                                       |
| platform                       | ios                                         |
| derive_catalyst_app_identifier | false                                       |
| fail_on_name_taken             | false                                       |
| skip_certificate_matching      | false                                       |
| skip_set_partition_list        | false                                       |
| verbose                        | false                                       |
+--------------------------------+---------------------------------------------+

[11:13:38]: Cloning remote git repo...
[11:13:38]: If cloning the repo takes too long, you can use the `clone_branch_directly` option in match.
[11:13:42]: Checking out branch viomsp-staging...
[11:13:42]: 🔓  Successfully decrypted certificates repo
[11:13:42]: Verifying that the certificate and profile are still valid on the Dev Portal...
Available session is not valid any more. Continuing with normal login.
[11:13:46]: Couldn't find a valid code signing identity for distribution... creating one for you now

+-------------------------+----------------------------------------------------------+
|                              Summary for cert 2.192.0                              |
+-------------------------+----------------------------------------------------------+
| platform                | ios                                                      |
| development             | false                                                    |
| generate_apple_certs    | false                                                    |
| force                   | true                                                     |
| username                | vineeth@fandemand.ca                                     |
| keychain_path           | /Users/vineethpalani/Library/Keychains/login.keychain-d  |
|                         | b                                                        |
| skip_set_partition_list | false                                                    |
+-------------------------+----------------------------------------------------------+

[!] Could not create another Distribution certificate, reached the maximum number of available Distribution certificates.

 </pre>
</details>

### Environment

<details><summary>🚫 fastlane environment 🚫</summary>

### Stack

| Key                         | Value                                                          |
| --------------------------- | -------------------------------------------------------------- |
| OS                          | 11.6                                                           |
| Ruby                        | 2.6.3                                                          |
| Bundler?                    | false                                                          |
| Git                         | git version 2.30.1 (Apple Git-130)                             |
| Installation Source         | /usr/local/bin/fastlane                                        |
| Host                        | macOS 11.6 (20G165)                                            |
| Ruby Lib Dir                | /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib |
| OpenSSL Version             | LibreSSL 2.8.3                                                 |
| Is contained                | false                                                          |
| Is homebrew                 | false                                                          |
| Is installed via Fabric.app | false                                                          |
| Xcode Path                  | /Applications/Xcode.app/Contents/Developer/                    |
| Xcode Version               | 13.0                                                           |
| Swift Version               | 5.5                                                            |

### System Locale

| Variable | Value       |   |
| -------- | ----------- | - |
| LANG     | en_US.UTF-8 | ✅ |
| LC_ALL   |             |   |
| LANGUAGE |             |   |

### fastlane files:

<details><summary>`./Fastfile`</summary>

```ruby
import "../../scripts/fastfile"

default_platform(:ios)

DEVELOPER_APP_ID = ENV["DEVELOPER_APP_ID"]
DEVELOPER_APP_IDENTIFIER = ENV["DEVELOPER_APP_IDENTIFIER"]
APPLE_KEY_ID = ENV["APPLE_KEY_ID"]
APPLE_ISSUER_ID = ENV["APPLE_ISSUER_ID"]
APPLE_API_KEY = ENV["APPLE_API_KEY"]
CHANGELOG = ENV["CHANGELOG"]
TESTER_GROUP = ENV["TESTER_GROUP"]
GIT_AUTHORIZATION = ENV["GIT_AUTHORIZATION"]
#PROVISIONING_PROFILE_SPECIFIER = ENV["PROVISIONING_PROFILE_SPECIFIER"]
TEMP_KEYCHAIN_USER = ENV["TEMP_KEYCHAIN_USER"]
TEMP_KEYCHAIN_PASSWORD = ENV["TEMP_KEYCHAIN_PASSWORD"]
FIREBASE_TOKEN = ENV['FIREBASE_TOKEN']
def delete_temp_keychain(name)
  delete_keychain(
    name: name
  ) if File.exist? File.expand_path("~/Library/Keychains/#{name}-db")
end

def create_temp_keychain(name, password)
  create_keychain(
    name: name,
    password: password,
    unlock: true,
    timeout: 0,
    default_keychain: true,
  )
end

def ensure_temp_keychain(name, password)
  delete_temp_keychain(name)
  create_temp_keychain(name, password)
end

platform :ios do
  lane :update_devices do
    keychain_name = TEMP_KEYCHAIN_USER
    keychain_password = TEMP_KEYCHAIN_PASSWORD
    ensure_temp_keychain(keychain_name, keychain_password)
    match(
      type: 'adhoc',
      app_identifier: "#{DEVELOPER_APP_IDENTIFIER}",
      git_private_key: "#{GIT_AUTHORIZATION}",
      readonly: true,
      keychain_name: keychain_name,
      keychain_password: keychain_password,
      force_for_new_devices: true
    ) 
    delete_temp_keychain(keychain_name)
  end
  # Updates XCode project settings to use a different code signing based on method
  private_lane :archive do |options|
    keychain_name = TEMP_KEYCHAIN_USER
    keychain_password = TEMP_KEYCHAIN_PASSWORD
    ensure_temp_keychain(keychain_name, keychain_password)
    method = options[:method]

    #profile_name = method == 'ad-hoc' ? "match AdHoc #{DEVELOPER_APP_IDENTIFIER}" : "Production"
    #case method
    #when 'development'
    profile_name = "#{ENV["MATCH_STRING"]}"
    #when 'ad-hoc'
    #  profile_name = "match AdHoc #{DEVELOPER_APP_IDENTIFIER}" 
    #when 'app-store'
    #  profile_name = "match AppStore #{DEVELOPER_APP_IDENTIFIER}"
    #else
    #  profile_name = "match AdHoc #{DEVELOPER_APP_IDENTIFIER}"
    #end
    update_code_signing_settings(profile_name: profile_name, bundle_identifier: "#{DEVELOPER_APP_IDENTIFIER}")

    case method
    when 'development'
      #register_devices(devices_file: "./devices.tsv")
      match(
        type: 'development',
        app_identifier: "#{DEVELOPER_APP_IDENTIFIER}",
        git_private_key: "#{GIT_AUTHORIZATION}",
        readonly: true,
        keychain_name: keychain_name,
        keychain_password: keychain_password
      ) 
    when 'ad-hoc'
      #register_devices(devices_file: "./devices.tsv")
      match(
        type: 'adhoc',
        app_identifier: "#{DEVELOPER_APP_IDENTIFIER}",
        git_private_key: "#{GIT_AUTHORIZATION}",
        readonly: true,
        keychain_name: keychain_name,
        keychain_password: keychain_password
      ) 
    when 'app-store'
      match(
        type: 'appstore',
        app_identifier: "#{DEVELOPER_APP_IDENTIFIER}",
        git_private_key: "#{GIT_AUTHORIZATION}",
        readonly: true,
        keychain_name: keychain_name,
        keychain_password: keychain_password 
      )
    else
      profile_name = "match AdHoc #{DEVELOPER_APP_IDENTIFIER}"
    end

    build_app(export_method: method)
    delete_temp_keychain(keychain_name)
  end

  private_lane :authenticate_apple_store do
    #p8_path = path_for_secret("your .p8 file path")

    app_store_connect_api_key(
      key_id: APPLE_KEY_ID,
      issuer_id: APPLE_ISSUER_ID,
      key_content: APPLE_API_KEY,
      duration: 1200, # optional
      in_house: false,
    )
  end

  lane :build do |options|
    # Reuse parent fastfile tasks
    fetch_dependencies
    #build_autogenerated_code


    sign_enabled = options[:sign_enabled] || false
    sign_param = sign_enabled ? '' : '--no-codesign'

    tree_shake_disabled = options[:no_tree_shake] || false
    tree_shake_param = tree_shake_disabled ? '--no-tree-shake-icons' : ''

    update_version = options[:update_version] || false
    if update_version
      sh_on_root(command: "flutter pub run  bin/maintain_app_version.dart -a #{ENV["APP_REF"]} -e #{ENV["ENV"]} -g repo -p ios -c #{ENV["COMMIT"]}")
    end
    sh_on_root(command: "flutter pub run tool/env.dart -a #{ENV["APP_REF"]} -e #{ENV["ENV"]} -n  #{ENV["APP_TITLE"]}")
    update_app_identifier(
      app_identifier: "#{DEVELOPER_APP_IDENTIFIER}",
      plist_path: "./Runner/Info.plist", # Path to info plist file, relative to xcodeproj
    )
    sh_on_root(command: "flutter build ios --release #{sign_param} #{tree_shake_param}")
  end

  lane :deploy_staging do
    build(sign_enabled: false, no_tree_shake: true, update_version: true)
    archive(method: "ad-hoc")

    #upload_symbols_to_crashlytics(dsym_path: "your.app.dSYM.zip")

    firebase_app_distribution(
      app: DEVELOPER_APP_ID,
      groups: "#{TESTER_GROUP}",
      #release_notes: "Lots of amazing new features to test out!",
      firebase_cli_token: FIREBASE_TOKEN,
    )
  end

  lane :deploy_staging_testflight do
    build(sign_enabled: false, update_version: true)
    archive(method: "app-store")

    authenticate_apple_store

    # Reuse parent fastfile tasks
    #test

    upload_to_testflight(
      ipa: "./Runner.ipa",
      reject_build_waiting_for_review: true,
      skip_waiting_for_build_processing: true,
      distribute_external: true,
      notify_external_testers: true,
      groups: TESTER_GROUP,
      changelog: CHANGELOG
    )
  end

  lane :deploy_production do
    sh_on_root(command: "sh scripts/cp_env.sh")

    # All certificates and .p8 file should be fine on runnning machine
    build(sign_enabled: true, config_only: true)
    archive(method: "app-store")
    authenticate_apple_store

    # Reuse parent fastfile tasks
    #test

    deliver(
      ipa: "your.ipa",
      skip_metadata: true,
      skip_screenshots: true,
      submit_for_review: false,
      force: false,
      automatic_release: false,
      submission_information: {
        add_id_info_serves_ads: false,
        add_id_info_uses_idfa: false,
        export_compliance_uses_encryption: false,
      },
      precheck_include_in_app_purchases: false,
    )
  end

end
```
</details>

<details><summary>`./Appfile`</summary>

```ruby
app_identifier(ENV["DEVELOPER_APP_IDENTIFIER"]) # The bundle identifier of your app
apple_id(ENV["MATCH_APPLE_ID"]) # Your Apple email address

itc_team_id(ENV["FASTLANE_ITC_TEAM_ID"]) # App Store Connect Team ID
team_id(ENV["FASTLANE_TEAM_ID"]) # Developer Portal Team ID
```
</details>

### fastlane gems

| Gem      | Version | Update-Status      |
| -------- | ------- | ------------------ |
| fastlane | 2.192.0 | 🚫 Update available |

### Loaded fastlane plugins:

| Plugin                                    | Version   | Update-Status      |
| ----------------------------------------- | --------- | ------------------ |
| fastlane-plugin-flutter_version           | undefined | 💥 Check failed     |
| fastlane-plugin-firebase_app_distribution | 0.2.9     | 🚫 Update available |

<details><summary><b>Loaded gems</b></summary>

| Gem                                       | Version      |
| ----------------------------------------- | ------------ |
| did_you_mean                              | 1.3.0        |
| atomos                                    | 0.1.3        |
| claide                                    | 1.0.3        |
| colored2                                  | 3.1.2        |
| nanaimo                                   | 0.3.0        |
| xcodeproj                                 | 1.19.0       |
| rouge                                     | 2.0.7        |
| xcpretty                                  | 0.3.0        |
| terminal-notifier                         | 2.0.0        |
| unicode-display_width                     | 1.7.0        |
| terminal-table                            | 1.8.0        |
| plist                                     | 3.6.0        |
| public_suffix                             | 4.0.6        |
| addressable                               | 2.8.0        |
| multipart-post                            | 2.0.0        |
| word_wrap                                 | 1.0.0        |
| optparse                                  | 0.1.1        |
| tty-screen                                | 0.8.1        |
| tty-cursor                                | 0.7.1        |
| tty-spinner                               | 0.9.3        |
| artifactory                               | 3.0.15       |
| babosa                                    | 1.0.4        |
| colored                                   | 1.2          |
| highline                                  | 2.0.3        |
| commander                                 | 4.6.0        |
| unf_ext                                   | 0.0.7.7      |
| unf                                       | 0.1.4        |
| domain_name                               | 0.5.20190701 |
| http-cookie                               | 1.0.4        |
| faraday-cookie_jar                        | 0.0.7        |
| faraday_middleware                        | 1.0.0        |
| fastimage                                 | 2.2.4        |
| gh_inspector                              | 1.1.3        |
| mini_magick                               | 4.11.0       |
| naturally                                 | 2.2.1        |
| rubyzip                                   | 2.3.0        |
| security                                  | 0.1.3        |
| xcpretty-travis-formatter                 | 1.0.1        |
| dotenv                                    | 2.7.6        |
| simctl                                    | 1.6.8        |
| jwt                                       | 2.2.3        |
| emoji_regex                               | 3.2.2        |
| aws-eventstream                           | 1.1.1        |
| aws-sigv4                                 | 1.2.3        |
| aws-sdk-kms                               | 1.44.0       |
| aws-sdk-s3                                | 1.96.1       |
| CFPropertyList                            | 3.0.3        |
| excon                                     | 0.83.0       |
| ruby2_keywords                            | 0.0.4        |
| faraday-net_http_persistent               | 1.1.0        |
| faraday-net_http                          | 1.0.1        |
| faraday-excon                             | 1.1.0        |
| faraday-em_synchrony                      | 1.0.0        |
| faraday-em_http                           | 1.0.0        |
| faraday                                   | 1.4.3        |
| json                                      | 2.5.1        |
| bundler                                   | 2.2.21       |
| webrick                                   | 1.7.0        |
| rexml                                     | 3.2.5        |
| httpclient                                | 2.8.3        |
| multi_json                                | 1.15.0       |
| signet                                    | 0.15.0       |
| os                                        | 1.1.1        |
| memoist                                   | 0.16.2       |
| googleauth                                | 0.16.2       |
| mini_mime                                 | 1.1.0        |
| retriable                                 | 3.1.2        |
| trailblazer-option                        | 0.1.1        |
| declarative                               | 0.0.20       |
| uber                                      | 0.1.0        |
| representable                             | 3.1.1        |
| google-apis-core                          | 0.4.0        |
| google-apis-playcustomapp_v1              | 0.4.0        |
| google-apis-androidpublisher_v3           | 0.7.0        |
| rake                                      | 13.0.3       |
| digest-crc                                | 0.6.3        |
| google-apis-storage_v1                    | 0.5.0        |
| google-apis-iamcredentials_v1             | 0.5.0        |
| google-cloud-errors                       | 1.1.0        |
| google-cloud-env                          | 1.5.0        |
| google-cloud-core                         | 1.6.0        |
| google-cloud-storage                      | 1.32.0       |
| aws-partitions                            | 1.472.0      |
| jmespath                                  | 1.4.0        |
| aws-sdk-core                              | 3.115.0      |
| forwardable                               | 1.2.0        |
| logger                                    | 1.3.0        |
| date                                      | 2.0.0        |
| stringio                                  | 0.0.2        |
| ipaddr                                    | 1.2.2        |
| openssl                                   | 2.1.2        |
| zlib                                      | 1.0.0        |
| ostruct                                   | 0.1.0        |
| strscan                                   | 1.0.0        |
| io-console                                | 0.4.7        |
| fileutils                                 | 1.1.0        |
| etc                                       | 1.0.1        |
| libxml-ruby                               | 3.2.1        |
| psych                                     | 3.1.0        |
| mutex_m                                   | 0.1.0        |
| fastlane-plugin-firebase_app_distribution | 0.2.9        |
</details>

*generated on:* **2021-11-12**
</details>

**Comment by vin-fandemand** [comment]
@crazymanish I raised it as an issue because I didn't get any community response. I think the match feature is something anyone using Fastlane for ios would use and surprised that no one has any solution or best practice for this problem.

**Reply by ihak** [reply]
I faced the same issue without generate_apple_certs flag but then when I used the command with flag setting false it worked just fine. I think you are providing the flag value in the wrong way. I used the following command:

`bundle exec fastlane match --generate_apple_certs false`

So the only difference I see is the way false value is provided to the flag.

**Reply by vin-fandemand** [reply]
Thanks @ihak , I tried with the false flag, seems to give the same error. Is there anything else I should be checking ?

**Reply by marfurt** [reply]
Hi @vin-fandemand , did you figure out how to make it work?

I encounter the same issue when I try to generate an AdHoc profile on an enterprise account which already has an iOS Distribution certificate previously created by Fastlane Match. The `--generate_apple_certs false` option doesn't make any difference.

**Comment by George045** [comment]
I had this issue couple of days ago and  I used the command with flag setting false it and it worked hope this will help you out.  
[skylight paycards](https://www.skylightpaycard.bid/)

**Comment by vin-fandemand** [comment]
No, I am still facing the same problem.

**Comment by marfurt** [comment]
The reason of the issue is that Fastlane Match places certificates for enterprise `adhoc` accounts under the `distribution` directory, which means that it wouldn't reuse existing distribution certificates placed under `enterprise`.

A change was implemented at some point, but reverted because of side-effects, see https://github.com/fastlane/fastlane/pull/12281.

I was able to make it work without re-generating a new profile by manually creating a `distribution` symlink that points to the existing `enterprise` directory within the match repository. With this change, Match finds the distribution certificate and is able to generate an AdHoc profile.

**Reply by vin-fandemand** [reply]
how can we make this fix? this is really frustrating because I am sure many developers are using this feature and most likely facing this issue. Or is there a different way to do the same thing?

**Comment by taraspasichnyk** [comment]
Have you tried nuking everything and recreating a certificates?

**Comment by vin-fandemand** [comment]
Doing nuke every time a certificate expires or when a new user joins is
ridiculous...
Please let me know if you have any queries.
Regards,
Vineeth

**Reply by bhat-ganesh** [reply]
Im having the exact issue. I was able to use `development` cert without issue. Now trying to use `appstore` and get this error. 
```
❯❯❯ fastlane match appstore --generate_apple_certs false
Couldn't find a valid code signing identity for distribution... creating one for you now
Could not create another Distribution certificate, reached the maximum number of available Distribution certificates.
```

Can we pls get more attention on this? What is the best course of action here?

**Comment by kimmy-wang** [comment]
same issue.

**Comment by triplef** [comment]
I’d do the following:

1. revoke all distribution certificates from the developer portal
2. delete all distribution certificates from your certificates repository, and commit + push the changes
3. run Match again to generate a new distribution cert

**Comment by vin-fandemand** [comment]
I tried to add the certificates manually to the GitHub repository and run
match so that I can bypass the buggy certificate generation discussed here.
It didn't work. The files generated by match in github has a particular
naming convention. Its some encoded string. Anyone knows how it is
generated? Anyone tried it out manually?

**Comment by vin-fandemand** [comment]
When you revoke all the profiles, will any of the apps using the old
certificates work ?

**Comment by AmirAlahmedy** [comment]
``` shell
 fastlane match --verbose <profile_type> --username <apple_developer_username> --team_id <team_id> --keychain_password <keychain_password> --git_branch <git_branch> -a <bundle_id>,<bundle_id.share> --git_url <git_url> --clone_branch_directly true --generate_apple_certs true --shallow_clone true
```
I think the above command will use the available certificates in your git storage without creating new ones while generating new profiles.

**Comment by DemonDG** [comment]
I face the same problem.

**Comment by faimin** [comment]
<img width="741" height="475" alt="image" src="https://github.com/user-attachments/assets/1b59175d-4e5b-4cb4-b19f-145f2e480a50" />

For standard individual or organization accounts, this limit is typically **3**. Enterprise accounts may have a different limit.

so the solution is revoke any unused distribution certificate, then recreate it.
